### PR TITLE
[ARCTIC-1025][FLINK] Fix data duplication even if there is a primary key table with upsert enabled

### LINKED ISSUE #1025: [Bug]: The same _file_offset in the Change Store causes data duplication even if there is a primary key table with upsert enabled
### What happened?

For tables with a primary key and the property 'write.upsert.enabled' set, when the same transaction id commits multiple changes to a primary key data, it can lead to data duplication. And the '_file_offset' of these duplicate data must be duplicated.
image

### Affects Versions

0.3.x

### What engines are you seeing the problem on?

Flink

### How to reproduce

1.Create an Arctic table with a primary key and start a Flink ingestion job
2.The same primary key is used for Insert and Update operations during a single transaction commit

### Relevant log output

_No response_

### Anything else

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by lklhdu**
FIX #1025 

## Why are the changes needed?

To make sure only one primary key in key table with upsert enabled, it's necessary to delete data firstly when update_after not followed by update_before.

It may lead to some unnecessary delete data if there are some data interspersed with other primary key data. e.g. K1-UB, K2-UB, K1-UA, K2-UB
## Brief change log

  - The flink 1.12 and 1.14, 1.15 versions are affected.

## How was this patch tested?
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: core/src/main/java/com/netease/arctic/io/writer/BaseTaskWriter.java [hunk 1]
```diff
@@ -123,6 +123,10 @@ public void close() throws IOException {
     writerHolder.close();
   }
 
+  public PrimaryKeyData getPrimaryKey() {
```
Comment (zhoujinsong) [inline core/src/main/java/com/netease/arctic/io/writer/BaseTaskWriter.java:126]: I'd recommend making this method to be `protected`.

**Review by zhoujinsong — APPROVED** [review]
LGTM.

File: flink/v1.12/flink/src/main/java/com/netease/arctic/flink/write/FlinkChangeTaskWriter.java [hunk 2]
```diff
@@ -76,4 +96,25 @@ protected ChangeAction action(RowData data) {
...
+      primaryKey.primaryKey(asStructLike(row));
+      PrimaryKeyData copyKey = primaryKey.copy();
+
+      if (RowKind.UPDATE_AFTER.equals(rowKind)) {
+        if (!hasUpdateBeforeKeys.contains(copyKey)) {
+          row.setRowKind(RowKind.INSERT);
+        }
+        hasUpdateBeforeKeys.remove(copyKey);
+      } else {
+        hasUpdateBeforeKeys.add(copyKey);
```
Comment (YesOrNo828) [inline flink/v1.12/flink/src/main/java/com/netease/arctic/flink/write/FlinkChangeTaskWriter.java:116]:
```suggestion
        PrimaryKeyData copyKey = primaryKey.copy();
        hasUpdateBeforeKeys.add(copyKey);
```
